Use ISO datetime format

JanErikHuehne opened the pull request:
### Short description
Adjusted API datetime field to match an ISO standard


### Proposed changes
- Remove date and time fields from event model, add datetime fields for start and end
- Use the event view to convert date and time fields of the event form to datetime fields of event model
- Adjust API datetime fields: Rename them to "last_updated" and let them return the datetime in ISO format
- For Event-API: Make use new event fields start and end to return ISO-datetime fields instead of start_time, start_date, end_time, end_date
- Adjust the input test data json and the expected test result json files to match the new input and output formats
### Resolved issues
Fixes: #1011

timoludwig commented on integreat_cms/cms/views/events/event_list_view.py:
```diff
@@ -103,16 +103,19 @@ def get(self, request, *args, **kwargs):
         chunk = request.GET.get("page")
         event_chunk = paginator.get_page(chunk)
-
+        for event in event_chunk:
+            setattr(event, "start_date", event.start.date())
+            setattr(event, "start_time", event.start.time())
+            setattr(event, "end_date", event.end.date())
+            setattr(event, "end_time", event.end.time())
```
```suggestion
```
This looks very ugly to me. Could we rather adapt the templates as well so they also use the new fields?

JanErikHuehne replied:
Also done.

timoludwig reviewed (commented):
Awesome, congratulations to your first PR and thanks a lot! :tada: :rocket:

First of all: I'm sorry I labeled this issue as "good first issue" - it was definitely not :wink:
I am all the more impressed that you have fought your way through it. :muscle: 

In general, everything seems to work as expected :+1:

I see two major points that should be fixed before merging this PR:
1. We need to migrate our existing event data, otherwise all start and end dates will be reset to the point in time when we install this update on our server.
2. We should also adapt the event list templates to use the same new fields instead of setting custom attributes.

If you have any troubles implementing these changes, don't hesitate to ask!

JanErikHuehne commented on integreat_cms/cms/forms/events/event_form.py:
```diff
@@ -43,19 +63,11 @@ class Meta:
         model = Event
         #: The fields of the model which should be handled by this form
         fields = [
```
I have implemented it this way, works fine. I am still not sure if we have to use HiddenInputs as widgets for the `start` and `end`. I have implemented it to be sure.

ulliholtgrave commented:
@digitalfabrik/app-team FYI We are currently implementing your required ISO format :)

JanErikHuehne commented on integreat_cms/cms/views/events/event_form_view.py:
```diff
@@ -212,7 +222,20 @@ def post(self, request, **kwargs):
+            )
+            event.save()
+            event_form.save_m2m()
+            event_translation_form.instance.event = event
```
Done

JanErikHuehne commented on integreat_cms/api/v3/events.py:
```diff
@@ -26,15 +26,21 @@ def transform_event(event, custom_date=None):
-        if custom_date
-        else event.end_date,
+        "start": start,
+        "end": end,
```
done

timoludwig commented on integreat_cms/api/v3/events.py:
```diff
@@ -26,15 +26,28 @@ def transform_event(event, custom_date=None):
-        "start_time": event.start_time,
-        "end_time": event.end_time,
         "recurrence_id": event.recurrence_rule.id if event.recurrence_rule else None,
         "timezone": settings.CURRENT_TIME_ZONE,
```
```suggestion
        "timezone": event.timezone,
```

timoludwig commented on integreat_cms/cms/views/events/event_form_view.py:
```diff
@@ -73,9 +72,15 @@ def get(self, request, *args, **kwargs):
         if event_instance:
             poi_instance = event_instance.location
+            initial = {
+                "start_date": event_instance.start.date(),
+                "start_time": event_instance.start.time(),
+                "end_date": event_instance.end.date(),
+                "end_time": event_instance.end.time(),
```
```suggestion
                "start_date": event_instance.start_local.date(),
                "start_time": event_instance.start_local.time(),
                "end_date": event_instance.end_local.date(),
                "end_time": event_instance.end_local.time(),
```

timoludwig reviewed (commented):
Amazing, thanks! :pray: 

I have to admit that I misunderstood how Django handles timezones :sweat_smile: 
Even if dates are given in timezone aware local times, Django converts them to UTC and does not keep track of the original timezone when writing them to the database. So every time we retrieve a date from the database, we have to convert it to the local time in the region's timezone before we either display it in the CMS or deliver it via the API.
This makes everything a bit more complicated, I hope I did not mess up my suggestions.

JanErikHuehne commented on integreat_cms/cms/models/events/event.py:
```diff
@@ -85,10 +85,8 @@ class Event(AbstractContentModel):
-    start_time = models.TimeField(blank=True, verbose_name=_("start time"))
-    end_date = models.DateField(verbose_name=_("end date"))
-    end_time = models.TimeField(blank=True, verbose_name=_("end time"))
+    start = models.DateTimeField(verbose_name="start", default=datetime.now)
+    end = models.DateTimeField(verbose_name="end", default=datetime.now)
```
1. Missed those thanks! 
2. Yes. But I agree, that your suggested work around in the data migration is better :+1:

JanErikHuehne commented on integreat_cms/cms/migrations/0033_event_datetime.py:
```diff
@@ -0,0 +1,76 @@
+            field=models.DateTimeField(
+                default=datetime.datetime.now, verbose_name="start"
+            ),
+        ),
```
Makes sense. I think this might effect the internal representation in the database and in what order it is handled. So I am not sure of the effect of it.

JanErikHuehne commented on integreat_cms/cms/migrations/0033_event_datetime.py:
```diff
@@ -0,0 +1,76 @@
+            ),
+        ),
+        migrations.RunPython(start_and_end_init),
+        migrations.RemoveField(
```
I like this workaround :+1:

JanErikHuehne commented on integreat_cms/cms/migrations/0033_event_datetime.py:
```diff
@@ -0,0 +1,76 @@
+    :param schema_editor: The database abstraction layer that creates actual SQL code
+    :type schema_editor: ~django.db.backends.base.schema.BaseDatabaseSchemaEditor
+    """
+    events = apps.get_model("cms", "Event")
+    for event in events.objects.all():
+        event.start = datetime.datetime.combine(event.start_date, event.start_time)
+        event.end = datetime.datetime.combine(event.end_date, event.end_time)
```
Timezones overall seem to be huge pain in the neck. I think your handling makes a lot of sense though. I think it might get a bit more complicated when we want to convert the timezones to the local times of the end-user when displaying them. I think this will need a lot of special treatment, especially when checking for all day events. We will have to be very careful with the logic and ordering of the conversions then, in order for it to function in the excepted way.

JanErikHuehne commented on integreat_cms/cms/migrations/0033_event_datetime.py:
```diff
@@ -0,0 +1,76 @@
+    for event in events.objects.all():
+        event.start = datetime.datetime.combine(event.start_date, event.start_time)
+        event.end = datetime.datetime.combine(event.end_date, event.end_time)
+
```
Okay, I didn´t know about this aspect yet. But it makes a lot of sense :+1:

JanErikHuehne commented on integreat_cms/cms/migrations/0033_event_datetime.py:
```diff
@@ -0,0 +1,76 @@
+                default=datetime.datetime.now, verbose_name="start"
+            ),
+        ),
+        migrations.RunPython(start_and_end_init),
```
@timoludwig  
So this has no effect in the normal migration? Since the fields exist when we call the script? But are they not deleted right after in any case? Maybe I´m wrong here. But from the logic standpoint I see it as: If the fields start_time etc. do not exist, create them and then delete them right after?

timoludwig replied:
I'm not sure whether I understand your comment correctly... Yes, the reversing code has no effect on the normal migration, but only because it is not executed, not because it has no side effects...
The first arguments of [RunPython()](https://docs.djangoproject.com/en/3.2/ref/migration-operations/#runpython) is called only when the migration is applied, and the second argument only when the migration is unapplied. When [reversing migrations](https://docs.djangoproject.com/en/3.2/topics/migrations/#reversing-migrations), the opposite operations are executed in the reverse order, so:
#### Applying this migration:
1. Create new `start` and `end` fields
2. Execute `start_and_end_init()` to populate new fields with the combined values of the old fields
3. Make `start` and `end` fields non-nullable
4. Remove old `start_date`, `start_time`, `end_date` and `end_time` fields
#### Reversing this migration:
1. Create old `start_date`, `start_time`, `end_date` and `end_time` fields
2. Make `start` and `end` fields nullable
3. Execute `start_and_end_reverse()` to populate the old fields with the split values of the new fields
4. Remove new `start` and `end` fields

JanErikHuehne replied:
Ah got it! I miss understand that the second argument of RunPython is only run when the migration is unapplied. That is actually a cool concept, I have to dive deeper into the django migrations at some point I guess :D

JanErikHuehne replied:
Added it!

JanErikHuehne commented on integreat_cms/cms/forms/events/event_form.py:
```diff
@@ -43,20 +63,17 @@ class Meta:
-            "start_time": forms.TimeInput(format="%H:%M", attrs={"type": "time"}),
-            "end_time": forms.TimeInput(format="%H:%M", attrs={"type": "time"}),
             "icon": IconWidget(),
+            # Not sure if they have to be set to HiddenInputs to work correct
+            "start": forms.widgets.HiddenInput(),
+            "end": forms.widgets.HiddenInput(),
```
Alright, good to know! :v:

JanErikHuehne commented:
I think that we also have to update the expected test results as well to match the timezone. I think that part is still missing, or?

JanErikHuehne commented on integreat_cms/cms/models/events/event.py:
```diff
@@ -149,7 +147,7 @@ def is_all_day(self):
+        return self.start.time() == time.min and self.end.time() == time.max.replace(
             second=0, microsecond=0
         )
 
```
Makes sense. LGTM!

timoludwig commented on integreat_cms/api/v3/events.py:
```diff
@@ -2,13 +2,13 @@
 from django.utils import timezone
 from django.utils.html import strip_tags
-
+from django.utils import timezone
```
```suggestion

```
`timezone` is already imported in line 9

JanErikHuehne replied:
Wait my IDE still marked it as missing. Maybe something was still loading, my bad :D

timoludwig commented on integreat_cms/cms/forms/events/event_filter_form.py:
```diff
@@ -94,9 +94,14 @@ def apply(self, events, region, language_slug):
-            events = events.filter_upcoming(
-                self.cleaned_data["date_from"] or date.min
-            ).filter(end_date__lte=self.cleaned_data["date_to"] or date.max)
+            tzinfo = zoneinfo.ZoneInfo(region.timezone)
```
I guess we need to `import zoneinfo` in this file.

JanErikHuehne replied:
Resolved

timoludwig commented on integreat_cms/cms/forms/events/event_form.py:
```diff
@@ -1,6 +1,6 @@
 import logging
 
```
```suggestion
import zoneinfo

```

JanErikHuehne replied:
Datetime did also still miss here. I added both

timoludwig commented on integreat_cms/cms/migrations/0033_event_datetime.py:
```diff
@@ -0,0 +1,108 @@
+import zoneinfo
+from datetime import datetime
+from django.db import migrations, models
+from django.utils import timezone
```
```suggestion
```
This import is now unused

timoludwig commented on integreat_cms/cms/models/events/event.py:
```diff
@@ -35,9 +35,9 @@ def filter_upcoming(self, from_date=None):
         :rtype: ~integreat_cms.cms.models.events.event.EventQuerySet
         """
-        from_date = from_date or date.today()
+        from_date = from_date or timezone.now().date()
```
`from django.utils import timezone` is missing at the top of the file

JanErikHuehne replied:
Done

timoludwig commented on integreat_cms/cms/models/events/event.py:
```diff
@@ -149,10 +147,43 @@ def is_all_day(self):
         :rtype: bool
         """
-        return self.start_time == time.min and self.end_time == time.max.replace(
+        return self.start_local.time() == time.min and self.end_local.time() == time.max.replace(
             second=0, microsecond=0
```
```suggestion
        return (
            self.start_local.time() == time.min
            and self.end_local.time() == time.max.replace(second=0, microsecond=0)
```
Our code formatting tool [black](https://github.com/digitalfabrik/integreat-cms/blob/develop/dev-tools/black.sh) wants to reformat these lines...

JanErikHuehne replied:
Already resolved by [c79e687](https://github.com/digitalfabrik/integreat-cms/pull/1642/commits/c79e6870b2025d6d81e2ff24dcb2a496891c0b04)

timoludwig reviewed (commented):
Awesome, thanks a lot! :muscle: 

Now there is not too much left to do I think. After you fixed the last issues and adapted the tests, could you clean up the git history and rebase the branch? After that, I will test everything extensively again :grin:

> Timezones overall seem to be huge pain in the neck

Yes, they give me quite a headache as well :sweat_smile: 

> I think that we also have to update the expected test results as well to match the timezone. I think that part is still missing, or?

Yes, I didn't make suggestions for this part because it's probably easier to just retrieve the endpoints and copy + paste the result into the files instead of replacing the dates manually...

JanErikHuehne commented:
> Now there is not too much left to do I think. After you fixed the last issues and adapted the tests, could you clean up the git history and rebase the branch? After that, I will test everything extensively again 

Yes sure. But I´m not sure when I will get to it since I am going to Munich today. Maybe I will find time and space in the train. In any other case, I should have time for it sometime next week

JanErikHuehne commented:
I completed the rebase and everything seems to work fine. I squashed the commits into 1 and added one slightly longer commit description. @timoludwig if you find something while re-checking everything, let me know. I added the change in the unreleased block inside the changelog, I hope this is fine.

timoludwig reviewed (approved):
Great, thank you so much! :pray: 
Works flawlessly :heart_eyes:
